### PR TITLE
fix incorrect scan position during bitmap index words scan

### PR DESCRIPTION
**Pull request by SmartKeyerror**
Since this PR #11377 has fixed bitmap index scan when concurrent insert update full 
bitmap page, which resolved concurrent read on bitmap index when there's insert running 
in the backend may cause the bitmap scan read wrong tid:

1. Query on bitmap: A query starts and reads all bitmap pages to `PAGE_FULL`, increases
the next tid to fetch, and releases the lock after reading each page.
2. Concurrent insert: insert a tid into `PAGE_FULL` cause expand compressed words to
new words, and rearrange words into `PAGE_NEXT`.
3. Query on bitmap: fetch `PAGE_NEXT` and expect the first tid in it should equal the
saved next tid. But actually `PAGE_NEXT` now contains words used to belong in PAGE_FULL.
This causes the real next tid less than the expected next tid. But our scan keeps increasing
the wrong tid. And then this leads to a wrong result.

The PR used `_bitmap_catchup_to_next_tid()` function to adjust `result->lastScanWordNo`
to the correct position if there has concurrent read/write and causes rearranging words into the 
next bitmap index page, it used the value of `words->firstTid` and `result->nextTid` to judge
whether rearranging words into the next bitmap index page happened or not, this is not entirely
right.

This is because `BMIterateResult` can only store 16*1024=16384 TIDs, but `BMBatchWords`
can save 3968 words by default, a words is 64bit, even in the worst case (all words not 
compressed) , `BMBatchWords` can hold 3968 * 64 = 253952 TIDs. So if there has a `PAGE_FULL`,
the `PAGE_FULL` must scan it more than one time, **but the value of BMBatchWords->firstTid will 
not be updated during each scan, it will only be updated when new bitmap index pages are read**. 

Therefore, in the absence of concurrent read/write, if we need to scan the same `BMBatchWords` 
multiple times, it will lead to the wrong scan position, resulting in wrong output results, as #13446.

To summarize, we just need to check for a rearranged condition when the new bitmap index page is 
read from disk, we should do nothing when we scan the same `BMBatchWords`.

### CODE REVIEW

**Comment by Aegeaner** [comment]
> `BMBatchWords`
> can save 3968 words by default

`maxNumOfWords` in a `BMBatchWords`'s calculation is quite complicated, can you show how come with this result?

**Comment by SmartKeyerror** [comment]
> > `BMBatchWords`
> > can save 3968 words by default
> 
> `maxNumOfWords` in a `BMBatchWords`'s calculation is quite complicated, can you show how come with this result?

```
/*
 * Approximately 4078 words per 8K page
 */
#define BM_MAX_NUM_OF_HRL_WORDS_PER_PAGE \
	((BLCKSZ - \
	MAXALIGN(sizeof(PageHeaderData)) - \
	MAXALIGN(sizeof(BMBitmapOpaqueData)))/sizeof(BM_HRL_WORD))

/* approx 255 */
#define BM_MAX_NUM_OF_HEADER_WORDS \
	(((BM_MAX_NUM_OF_HRL_WORDS_PER_PAGE-1)/BM_HRL_WORD_SIZE) + 1)

/*
 * To make the last header word a complete word, we limit this number to
 * the multiplication of the word size.
 */
#define BM_NUM_OF_HRL_WORDS_PER_PAGE \
	(((BM_MAX_NUM_OF_HRL_WORDS_PER_PAGE - \
	  BM_MAX_NUM_OF_HEADER_WORDS)/BM_HRL_WORD_SIZE) * BM_HRL_WORD_SIZE)

#define BM_NUM_OF_HEADER_WORDS \
	(((BM_NUM_OF_HRL_WORDS_PER_PAGE-1)/BM_HRL_WORD_SIZE) + 1)
```

I make a simple diagram to illustrate this: 

image

A bitmap index page include 4 parts: 
- PageHeaderData
- BMBitmapOpaqueData
- Bitmap Index Header Section
- Bitmap Index Content Section

Since we used HRL compressed method to compress bitmaps, so we must have header section to indicate which word has been compressed in the content section. At the same time, whether it is header or content words, their size is 64 bits, equal to `BM_HRL_WORD_SIZE` and `sizeof(BM_HRL_WORD)`. **And one bit in the header will represent a word in the content section**.

Then, according to the above information, calculate the number of header words and the number of content words that a page can save, as `BM_NUM_OF_HEADER_WORDS` and `BM_NUM_OF_HRL_WORDS_PER_PAGE`.

**Comment by interma** [comment]
Thanks Zhenglong, at first glance. I think your PR is ok.

But historically, there are some problems with the bitmap index, especially in concurrent access.
I will take more time to think (and learn) the whole picture.

no hurry, :-)

**Review by interma — COMMENTED** [review]
Just small suggestion.
I am going to check other corner case later.

File: src/backend/access/bitmap/bitmaputil.c [hunk 1]
```diff
@@ -368,11 +364,20 @@ _bitmap_findnexttids(BMBatchWords *words, BMIterateResult *result,
...
 	if (words->firstTid >= result->nextTid)
 		return;
 
+	/*
+	 * if result->lastScanWordNo is equal to words->startNo, a new batch words
+	 * will be geneerated from next_batch_words(), and at this time if words->firstTid
+	 * is less than result->nextTid, then we should adjust result->lastScanWordNo to
+	 * the correct position.
+	 */
+	if (result->lastScanWordNo != words->startNo && !isStream)
```
[the remark this answers is not in the document]
Comment (interma) [reply src/backend/access/bitmap/bitmaputil.c:378]: if "just loaded a new page" and "words->firstTid > result->nextTid" 
then need catch up, right?

**Review by interma — APPROVED** [review]
LGTM

And last suggestion:
Considering a very simple index scan query can trigger this bug (but we haven't noticed it for a long time), do we need to add more additional test cases for bitmap index?

should cover:
* index scan, bitmap index scan, index only scan
* concurrent insert + select/insert
* bitmap operations, like union
* other corner cases

**Comment by SmartKeyerror** [comment]
> LGTM
> 
> And last suggestion: Considering a very simple index scan query can trigger this bug (but we haven't noticed it for a long time), do we need to add more additional test cases for bitmap index?
> 
> should cover:
> 
> * index scan, bitmap index scan, index only scan
> * concurrent insert + select/insert
> * bitmap operations, like union
> * other corner cases

Thanks for your review again! Absolutely we should add more tests on Bitmap Index, besides this issue will happen at a very simple and common scenario, there has another issue about storage: #13517.

Therefore, I think the problems found so far are just the tip of the iceberg, and we need to construct more complex test scenarios to find more problems.

This can take some time, so I prefer to solve the current problem first, and then focus on writing the test, do you think it's a good idea?

File: src/backend/access/bitmap/bitmaputil.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (SmartKeyerror) [reply src/backend/access/bitmap/bitmaputil.c:378]: > if "just loaded a new page" and "words->firstTid > result->nextTid" then need catch up, right?

Only in the situation that there have concurrent read/write on two adjacent pages, and inserting a tid into `PAGE_FULL` cause expand compressed words to new words, and rearrange words into `PAGE_NEXT`, and we ready to read a new page, we should catch up the `result-> lastScanWordNo` to the current position.

But this PR #11377 change a lot of things, even in the `pull_stream`, we need to catch up `result-> lastScanWordNo`, I think it's a side effect of adding `words->firstTid` to resolve the previous issue. Because in bitmap index scan, the first call will read all bitmap pages into memory.

File: src/backend/access/bitmap/bitmaputil.c [hunk 2]
```diff
@@ -368,11 +364,20 @@ _bitmap_findnexttids(BMBatchWords *words, BMIterateResult *result,
...
 void
-_bitmap_catchup_to_next_tid(BMBatchWords *words, BMIterateResult *result)
+_bitmap_catchup_to_next_tid(BMBatchWords *words, BMIterateResult *result, bool isStream)
 {
 	if (words->firstTid >= result->nextTid)
 		return;
 
+	/*
+	 * if result->lastScanWordNo is equal to words->startNo, a new batch words
+	 * will be geneerated from next_batch_words(), and at this time if words->firstTid
```
[the remark this answers is not in the document]
Comment (SmartKeyerror) [reply src/backend/access/bitmap/bitmaputil.c:374]: Nice catch, thx~

**Comment by interma** [comment]
> > LGTM
> > And last suggestion: Considering a very simple index scan query can trigger this bug (but we haven't noticed it for a long time), do we need to add more additional test cases for bitmap index?
> > should cover:
> > 
> > * index scan, bitmap index scan, index only scan
> > * concurrent insert + select/insert
> > * bitmap operations, like union
> > * other corner cases
> 
> Thanks for your review again! Absolutely we should add more tests on Bitmap Index, besides this issue will happen at a very simple and common scenario, there has another issue about storage: #13517.
> 
> Therefore, I think the problems found so far are just the tip of the iceberg, and we need to construct more complex test scenarios to find more problems.
> 
> This can take some time, so I prefer to solve the current problem first, and then focus on writing the test, do you think it's a good idea?

no problem~ thanks zhenglong

**Comment by SmartKeyerror** [comment]
@interma Hi~ I refactored some codes and made them more intelligible, could you have a look again? Thanks.

The pipeline failed in a flaky case, my local full pipeline does not fail in this case.

**Review by interma — APPROVED** [review]
LGTM

**Comment by SmartKeyerror** [comment]
The pipeline failed in an unrelated case, my local pipeline of ICW is green, so merge this.
